### PR TITLE
API-21315: FhirTokenParameter processes token parameters

### PR DESCRIPTION
**Pull request by jabando-va**

### CODE REVIEW

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package gov.va.api.health.fhir.api;
+
+import lombok.Value;
+
+@Value
+public class FhirTokenParameter {
```
Comment (jhulbert-va) [inline fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:6]: I think this doesn't do enough. What if I want to build a token parameter? I think it should follow our regular format. Can be parsed, and can be created.

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    return FhirTokenParameter.builder().system(parts[0]).code(parts[1]).build();
+  }
+
+  public boolean systemExistsIn(String... systems) {
+    return Arrays.stream(systems).anyMatch(s -> s.equals(system()));
+  }
+
+  @Override
+  public String toString() {
+    return system() + "|" + code();
```
Comment (jblefkove-va) [inline fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:32]: Should this check if `system` is null?

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  public static final class InvalidRequest extends IllegalArgumentException {
+    /** Invalid request exception. */
+    public InvalidRequest(String value) {
+      super(
+          String.format(
+              "invalid fhir token %s, expected format to be {system}|{code} or {code}", value));
```
Comment (jblefkove-va) [inline fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:40]: Is this class intentionally excluding `{system}|` and `|{code}` ?

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
+package gov.va.api.health.fhir.api;
+
+import java.util.Arrays;
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder
+public class FhirTokenParameter {
```
Comment (jblefkove-va) [inline fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:9]: What is the distinction between this class and Vulcan [TokenParameter](https://github.com/department-of-veterans-affairs/lighthouse-vulcan/blob/main/src/main/java/gov/va/api/lighthouse/vulcan/mappings/TokenParameter.java)?

Comment (jabando-va) [reply fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:9]: Ultimately they have the same roles, but we wanted Vulcan's use to be limited to applications with database-query processing like data-query/CDW. I think we decided that we did not want to use VFQ and CFQ with Vulcan, but we wanted to have some similar functionality available to them.

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+  String originalString;
```
Comment (jhulbert-va) [inline fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:13]: Not a fan of using string in the variable name for a string. `original` or `originalValue` would be better.

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+  String system;
+
+  String code;
+
+  /** Extract system (if present) and code from parameter string. */
+  public static FhirTokenParameter parseToken(String paramString) {
+    var parts = paramString.split("\\|", -1);
+    if ((parts.length != 1 && parts.length != 2) || paramString.equals("|")) {
```
Comment (jhulbert-va) [inline fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:22]: I think there should be an `isBlank` check before splitting. Then you would only need to check that the param string is not `|` and that the array length is not more than 2.

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      return FhirTokenParameter.builder().originalString(paramString).system(parts[0]).build();
+    }
+    return FhirTokenParameter.builder()
+        .originalString(paramString)
+        .system(parts[0])
+        .code(parts[1])
+        .build();
+  }
+
+  public boolean systemExistsIn(String... systems) {
```
Comment (jhulbert-va) [inline fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:41]: `isSystemOneOf`?

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    return Arrays.stream(systems).anyMatch(s -> s.equals(system()));
+  }
+
+  @Override
+  public String toString() {
+    return originalString();
+  }
```
Comment (jhulbert-va) [inline fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:48]:
```
FhirTokenParameter.builder().code("abc").build().toString()
```

^ That'll return null, but it should return `abc`. I'm all for storing the originalString, but we cannot assume we'll always have it.

**Review by jblefkove-va — APPROVED** [review]

**Pull request merged**